### DISCUSSION TITLE
What is the Staging equivalent of `SDL_VIDEODRIVER=dummy`

### DISCUSSION THREAD

**Discussion by LowLevelMahn** [dosbox-staging/dosbox-staging#3985] [Q&A]
i try to prevent the window opening but `SDL_VIDEODRIVER=dummy` does not work with dosbox-staging/SDL2

for example i want to use dosbox-staging in a docker image without gui or WSL without ui extension - so no ui possible

tried `SDL_VIDEODRIVER=offscreen' but that does only work with Linux in a VM but not with Windows or Linux in Windows Subsystem for Linux (WSL) - and especially with WSL the per default fullscreen mode is very annoying when several scripts gets runned by dosbox in the background and the screen flips like hell from starting/stoping dosbox

Windows:
![image](https://github.com/user-attachments/assets/34f77aa5-8120-4261-8e66-516f49868465)

Linux/WSL
![image](https://github.com/user-attachments/assets/4f0ed894-a865-4b21-9f4e-023c7138e203)

is there an option for console only so no graphical output?

**Comment by johnnovak** [comment]
> is there an option for console only so no graphical output?

No.

This is an extremely niche ask, so I guess you'll either need to live with the screen flickering (because your build scripts work that way, right?), or PR a change yourself. But unless the change will be very straightforward, we don't wanna add complication/maintenance burden for "headless mode".

An emulator meant for playing DOS games does not need a "headless mode", it makes no sense. It conflicts with the project goals, so we'll only do this if it's a few lines of trivial code.

**Reply by LowLevelMahn** [reply]
> This is an extremely niche ask

not if you're doing project with the need of using old compilers :) - so more or less all dos game reversing projects - but still a niche :)

> so we'll only do this if it's a few lines of trivial code.

i think dosbox vanilla doesn't did something different here - it was only a side-effect of SDL1 that is mostly used here
i try to analyse it a litte an maybe come up with some trivial lines of code you all like

the Open Watcom v2 maintainer tried to not use `set SDL_VIDEODRIVER=dummy` with the build CI but failed

> I am sorry but I must return DOSBOX setup back because we need to run DOSBOX on background without terminal.
> It is reason for SDL_VIDEODRIVER=dummy that DOSBOX not need terminal any way.
> As soon as some driver is used then it requires interactive session that doesn't exists.

**Reply by johnnovak** [reply]
> > This is an extremely niche ask
> 
> not if you're doing project with the need of using old compilers :) - so more or less all dos game reversing projects - but still a niche :)

Which _is_ a niche ask—you've just spelt it out in detail, man 😆  Read about our project goals here:

https://www.dosbox-staging.org/about/

It's all about the games, then DOS demos and apps on a best-effort basis. Interacting with scripts in headless mode for dev-only workflows like this one is simply not in our list.

My suggestion is to drop DOSBox and start using some other tool like PCem, 86box, or whatever.

Or just drop this fully automated tooling idea, and execute whatever you need in DOSBox manually on demand. Like how often do you recompile the compiler itself, really 😆 When you do, just live with the blinking windows for those 5 minutes 🤷🏻  1 out of 100,000 people need this to be _fully automated_ in headless mode...

Then again, PRs welcome (if it's simple to make this work) 😄

**Reply by LowLevelMahn** [reply]
> Which _is_ a niche ask—you've just spelt it out in detail

i know its a niche - never said anything else - absolutely understand your point

> My suggestion is to drop DOSBox and start using some other tool like PCem, 86box, or whatever.

the projects will just use dosbox(or dosbox-x) as before (for many years) no such big changes needed :)

> When you do, just live with the blinking windows for those 5 minutes

if you got the possibility to open up a window at all - whats not typical in docker or WSL builds - like with the github build-server for Open Watcom v2 - but as i said using dosbox classic or dosbox-x is still ok

i only stumbled over this because Fedora/Tumbleweed switched to dosbox staging a while ago and all my typical headless dosbox scenarios just didn't worked anymore - thought it was a bug - buts its intended and understandable

hope to come up with a small patch for this but could be not that easy to be just trivial :)

**Reply by johnnovak** [reply]
@LowLevelMahn Btw, if the mouse is the problem, why don't you just disable it in your build workflow and be done with it?

Try these:

```ini
[mouse]
dos_mouse_driver = off

# you might wanna try these too if the above alone doesn't do the trick
#mouse_raw_input = off
#mouse_multi_display_aware = off
#mouse_capture = nomouse
#ps2_mouse_model = none
```

Then if the "put mouse into relative mode" call still happens with the internal mouse emulation disabled, a small PR that fixes that is much more likely to get merged 😉

**Reply by LowLevelMahn** [reply]
> @LowLevelMahn Btw, if the mouse is the problem, why don't you just disable it in your build workflow and be done with it?
> 
> Try these:
> 
> ```ini
> [mouse]
> dos_mouse_driver = off
> 
> # you might wanna try these too if the above alone doesn't do the trick
> #mouse_raw_input = off
> #mouse_multi_display_aware = off
> #mouse_capture = nomouse
> #ps2_mouse_model = none
> ```
> 
> Then if the "put mouse into relative mode" call still happens with the internal mouse emulation disabled, a small PR that fixes that is much more likely to get merged 😉

that is only the error when using SDL_VIDEODRIVER=offscreen on Linux/WSL, Windows fails different and just don't get the offscreen driver - it behaves different on every systems

the former SDL1 SDL_VIDEODRIVER=dummy worked on all platforms (Linux, Linux/WSL, Windows) the same way with dosbox classic

im currently writing an easy step byte step howto build Open Watcom v2 on all supported platforms - and try to fix the makefiles to works the same with dosbox-staging on the github build-server (NEEDS headless) and Linux (Ubuntu, Fedora, Tumbleweed), Linux on WSL and Windows - im just trying to find an easy solution as with dosbox classic or dosbox-x - but using the better dosbox-staging :) - even if that means that the server needs to use dosbox class/x or telling the users to patch here and there in the makefiles before build (if they want to use dosbox-staging)

**Comment by interloper98** [comment]
Check out `xvfb-run`. It's how X11 program testing is done headless in CI. The arg line would be something like: `xvfb-run dosbox [dosbox args] -c exit`.

You don't need to fake out SDL with environment variables -- it runs it a real X11 session. The virtual X11 session will exit when the program exits and forward the return code back to your shell.

**Reply by LowLevelMahn** [reply]
is there also an solution for Windows? (the xvfb-run could maybe also work with Linux in WSL)

**Reply by interloper98** [reply]
Hmm; looks like the solutions with Windows are a lot clunkier (like using a VM or something heavy like that). 

In that case, using the SDL environment variables is probably the easiest.

I would suggest that the upstream project yield to existing SDL environment variables (and only set them if they don't exist), so users can set their own values.

**Reply by LowLevelMahn** [reply]
> In that case, using the SDL environment variables is probably the easiest.

thats the case - its does work with dosbox classic but not with dosbox-staging - i think because of SDL2 or because dosbox staging is doing more - like trying to "put mouse in relative mode" - things that dosbox classic just don't do

**Comment by grapeli** [comment]
Fortunately, there is still a possibility to use this completely useful and easy to run functionality.

In case of using the "`dummy`" driver, it is necessary to compile dosbox-staging with the option "`-Duse_opengl=false`".

`SDL_VIDEODRIVER=dummy dosbox -set output=texture -set texture_renderer=software -c exit`

In case of the "`offscreen`" driver, the mouse capture option should be disabled.

`SDL_VIDEODRIVER=offscreen dosbox -set output=texture -set mouse_capture=nomouse -c exit`

**Reply by LowLevelMahn** [reply]
thanks, will try that on the platforms

maybe you know that: do your see a chance to have that feature at runtime - based on a "small patch johnnovak would accept" :)
or is there a huge amount of code to change to make it runtime-able - without using `-Duse_opengl=false` on compilation

**Reply by FeralChild64** [reply]
I can only get rid of the _Failed to put the mouse in relative-mode_ forced exit by turning it into a warning and ignoring the problem.

**Comment by grapeli** [comment]
The dosbox-staging (with opengl) c6722e3 version works fine with dummy. You'd have to check at which change it stopped working.

edit: It stopped working after this change e1f7519 "Fix OpenGL attribute initialization".

**Comment by johnnovak** [comment]
Closing, _way_ out of scope.

**Comment by Torinde** [comment]
Additional usecase:
- XCOPY build from source https://github.com/dosbox-staging/dosbox-staging/issues/4447#issuecomment-3161897090

**Reply by johnnovak** [reply]
Just nope. It's crazy to build stuff like that on CI that changes once every 5 years. At best.

**Reply by LowLevelMahn** [reply]
...and it can be build for dos using a windows/linux DOS compiler like Watcom

**Reply by laser-eagle** [reply]
My usecase my interest you, I'm using this to run old 16bit software for an environmental science lab that monitors waste treatment. I can live with the popup, but it is something I'd like an option to disable.

**Reply by johnnovak** [reply]
> My usecase my interest you, I'm using this to run old 16bit software for an environmental science lab that monitors waste treatment. I can live with the popup, but it is something I'd like an option to disable.

Read our goals on our about page. Such usage is out of scope for us and it's heavily discouraged.
